dace/codegen/compiled_sdfg.py

```diff
@@ -407,23 +453,40 @@ def __del__(self):
         self._lib.unload()
 
     def _construct_args(self, kwargs) -> Tuple[Tuple[Any], Tuple[Any]]:
-        """ Main function that controls argument construction for calling
-            the C prototype of the SDFG.
+        """
+        Main function that controls argument construction for calling
+        the C prototype of the SDFG.
+
+        Organizes arguments first by ``sdfg.arglist``, then data descriptors
+        by alphabetical order, then symbols by alphabetical order.
 
-            Organizes arguments first by `sdfg.arglist`, then data descriptors
-            by alphabetical order, then symbols by alphabetical order.
+        :note: If not initialized this function will initialize the memory for
+               the return values, however, it might also reallocate said memory.
+        :note: This function will also update the internal argument cache.
         """
-        # Return value initialization (for values that have not been given)
         self._initialize_return_values(kwargs)
+
+        # Add the return values to the arguments, since they are part of the C signature.
         for desc, arr in zip(self._retarray_shapes, self._return_arrays):
             kwargs[desc[0]] = arr
 
-        # Argument construction
+        arglist, argtypes, argnames = self._construct_args_arglist(kwargs)
```

Comment:
Function calls can be expensive, performance-wise.
If you think they are not, please show some runtime results that say otherwise.

Reply:
As requested I performed some experiment.
I used the attached function, which is based on the `spmv` tutorial (randomly selected, if I should test another function please provide one).

I got the following results for the current master (`09d37e9e`):
```
min: 0.0001241610007127747s
max: 0.0007202319975476712s
mean: 0.0001336343956439426s
std: 3.586189058597028e-05
q_{1, 5, 25, 50, 75, 95, 99}: [0.00012455 0.00012489 0.00012555 0.00012643 0.00012851 0.00014832 0.00037179]
```

For my previous state, denoted append version, (`eb0198a5`) I got the following results:
```
min: 0.00010327700147172436s
max: 0.0006421329999284353s
mean: 0.00011245609261338056s
std: 3.2908162087660314e-05
q_{1, 5, 25, 50, 75, 95, 99}: [0.00010416 0.00010479 0.0001056 0.00010633 0.00010738 0.00012325 0.00030589]
```

And for the current version (`ff3ced35`):
```
min: 0.00010886099698836915s
max: 0.0006989000030444004s
mean: 0.00011409736964318047s
std: 2.089365416625181e-05
q_{1, 5, 25, 50, 75, 95, 99}: [0.00010925 0.00010962 0.00011023 0.000111 0.00011243 0.00011744 0.00020909]
```

As you can see from these results even the append version is faster than the current master.
According to my memory the majority of these gains come from caching some type checking results.
We also see that the new version is a little bit slower (in case of `min`) than the append version, but the differences are very small, but its 99 percentile is much lower.
I think this should suffice to show that this version is not slower than the original version.
[call_args.py.gz](https://github.com/spcl/dace/files/13776544/call_args.py.gz)

Reply:
Because you insisted so much on it I decided to put everything back into one function (`1a1fca4b`).
However, I also run the test again with the following results:
```
min: 0.00010630900214891881s
max: 0.0006737619987688959s
mean: 0.00011147427259614536s
std: 2.1202032309282628e-05
q_{1, 5, 25, 50, 75, 95, 99}: [0.00010679 0.00010708 0.00010762 0.00010819 0.00010998 0.00011565 0.00020852]
```
As you can see there is not much difference compared to the previous (`ff3ced35`) version where the `_construct_args()` function was separated into smaller pieces.